C/tasks/statistics_history/stats_history.cpp

```diff
@@ -59,33 +66,72 @@ void StatsHistory::run() const
 	ResultSet *keySet = getStorageClient()->queryTable("statistics", query);
 
 	ResultSet::RowIterator rowIter = keySet->firstRow();
+	std::vector<InsertValues> historyValues;
+	vector<pair<InsertValue *, Where *>> updateValues;
 
-        do {
+        while (keySet->hasNextRow(rowIter) || keySet->isLastRow(rowIter) )
+	{
 		string key = (*rowIter)->getColumn("key")->getString();
 		try {
-			processKey(key);
+			processKey(key, historyValues, updateValues);
 		} catch (exception e) {
 			getLogger()->error("Failed to process statisitics key %s, %s", key, e.what());
 		}
-                rowIter = keySet->nextRow(rowIter);
-	} while (keySet->hasNextRow(rowIter));
+		if (!keySet->isLastRow(rowIter))
+                	rowIter = keySet->nextRow(rowIter);
+		else
+			break;
+	}
+
+	int n_rows;
+
+        if ((n_rows = getStorageClient()->insertTable("statistics_history", historyValues)) < 1)
+        {
+                getLogger()->error("Failed to insert rows to statistics history table ");
+        }
+
+	if (getStorageClient()->updateTable("statistics", updateValues) < 1)
+        {
+                getLogger()->error("Failed to update rows to statistics table");
+        }
+
+	for (auto it = updateValues.begin(); it != updateValues.end() ; ++it)
+	{
+		InsertValue *updateValue = it->first;
+		if (updateValue)
+		{
+			delete updateValue;
+			updateValue=nullptr;
+		}
+        	Where *wKey = it->second;
+		if(wKey)
+		{
+			delete wKey;
+			wKey = nullptr;
+		}
+	}
 
 	delete keySet;
+
 }
 
 /**
- * Process a single statistics key
+ * Process statistics keys
  *
- * @param key	The statistics key to process
+ * @param key	         The statistics key to process
+ * @param historyValues  Values to be inserted in statistics_history
+ * @param updateValues   Values to be updated in statistics
+ * @return void
  */
-void StatsHistory::processKey(const string& key) const
+void StatsHistory::processKey(const std::string& key, std::vector<InsertValues> &historyValues, std::vector<std::pair<InsertValue*, Where *> > &updateValues) const
 {
-Query	query(new Where("key", Equals, key));
+	Query	query(new Where("key", Equals, key));
```

Comment:
We should not query the statistics table in a loop for each key. Get all stats, find delta for each, put into stats history as a batch insert with same history_ts

Reply:
fixed